File: revolut/ExchangeViewController.m
//
//  ExchangeViewController.m
//  revolut
//
//  Created by Alexander Danilov on 31/08/2017.
//  Copyright © 2017 Home. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "ExchangeViewController.h"
#import "ExchangeAccountPickerView.h"
#import "ExchangeViewModel.h"
#import "AppDelegate.h"

@interface ExchangeViewController ()
    
@property (weak, nonatomic) IBOutlet UIBarButtonItem *exchangeButton;
    
@end

@implementation ExchangeViewController {
    ExchangeViewModel *_viewModel;
}

-(void)awakeFromNib {
    [super awakeFromNib];
    _viewModel = [AppDelegate sharedInstance].exchangeViewModel;
}

-(void)exchange {
    NSError *error;
    if (![_viewModel exchange:&error]) {
        [[AppDelegate sharedInstance] onError:error];
    }
}

#pragma mark - UIViewController

-(void)viewDidLoad {
    [super viewDidLoad];
    [_exchangeButton setTarget:self];
    [_exchangeButton setAction:@selector(exchange)];
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [_viewModel addObserver:self forKeyPath:@"canExchange" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionInitial context:nil];
}

-(void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    
    [_viewModel removeObserver:self forKeyPath:@"canExchange"];
}

#pragma mark - UITableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 2;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ExchangeTableCell" forIndexPath:indexPath];
    
    UIPageControl *pageControl = [[cell.contentView subviews] objectAtIndex:1];
    
    ExchangeAccountPickerView *accountPicker = [[cell.contentView subviews] objectAtIndex:0];
    accountPicker.backgroundColor = [accountPicker.backgroundColor colorWithAlphaComponent:indexPath.row == 0 ? 0.7 : 1];
    accountPicker.pageControl = pageControl;
    [accountPicker setViewModel:_viewModel andType: indexPath.row == 0 ? kAccountPickerFrom : kAccountPickerTo];
    
    return cell;
}

#pragma mark - NSObject

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if ([keyPath isEqualToString:@"canExchange"]) { // TODO add and check observeContext
        _exchangeButton.enabled = _viewModel.canExchange;
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

@end

File: revolut/ExchangeViewController.swift
//
//  ExchangeViewController.swift
//  revolut
//
//  Created by Alexander Danilov on 12/08/2018.
//  Copyright © 2018 Home. All rights reserved.
//

import UIKit
import Foundation

class ExchangeViewController: UITableViewController {
    @IBOutlet weak var exchangeButton: UIBarButtonItem!
    
    private let viewModel: ExchangeViewModel
    
    required init?(coder aDecoder: NSCoder) {
        viewModel = AppDelegate.shared.exchangeViewModel
        super.init(coder: aDecoder)
    }
    
    func exchange() {
        do {
            try viewModel.exchange()
        } catch let error {
            AppDelegate.shared.onError(error: error)
        }
    }
    
    // MARK: - UIViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        exchangeButton.target = self
        exchangeButton.action = #selector(exchange)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.viewModel.addObserver(self, forKeyPath: "canExchange", options: [.new, .initial], context: nil)
    }

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        self.viewModel.removeObserver(self, forKeyPath: "canExchange")
    }
    
    // MARK: - UITableView
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 2
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ExchangeTableCell", for: indexPath)
        
        let pageControl = cell.contentView.subviews[1] as! UIPageControl
        
        let accountPicker = cell.contentView.subviews[0] as! ExchangeAccountPickerView
        accountPicker.backgroundColor = accountPicker.backgroundColor?.withAlphaComponent(indexPath.row == 0 ? 0.7 : 1)
        accountPicker.pageControl = pageControl
        accountPicker.setViewModel(viewModel, andType: indexPath.row == 0 ? .from : .to)
        
        return cell
    }
    
    // MARK: - NSObject
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if (keyPath == "canExchange") { // TODO add and check observeContext
            exchangeButton.isEnabled = viewModel.canExchange
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
}
